### PR TITLE
Reset a failed tool call

### PR DESCRIPTION
**Pull request by enyst**
**End-user friendly description of the problem this fixes or functionality that this introduces**

- [ ] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

This PR proposes a quick fix to:
- add an ErrorObs for a tool call pending when the agent is STOPPED or in ERROR. Otherwise these actions will not be in agent's history sent to the model
- fix log tool calls in responses

---
**Link of any specific issues this addresses**

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:e8bd498-nikolaik   --name openhands-app-e8bd498   docker.all-hands.dev/all-hands-ai/openhands:e8bd498
```

### CODE REVIEW

File: openhands/controller/agent_controller.py [hunk 1]
```diff
@@ -334,6 +334,10 @@ async def _handle_message_action(self, action: MessageAction) -> None:
     def _reset(self) -> None:
         """Resets the agent controller"""
 
+        # if the pending action has a tool call, reset the tool call
+        if self._pending_action and hasattr(self._pending_action, 'tool_call_metadata'):
+            self._pending_action.tool_call_metadata = None
```
Comment (xingyaoww) [inline openhands/controller/agent_controller.py:339]: Any idea why we want to do this? Why not directly remove the pending_action?

File: openhands/agenthub/codeact_agent/codeact_agent.py [hunk 2]
```diff
@@ -482,7 +482,14 @@ def _get_messages(self, state: State) -> list[Message]:
...
                         and message.role != 'tool'
                     ):
-                        messages[-1].content.extend(message.content)
+                        # find the last TextContent object in the last message
+                        last_text_content = None
+                        for content in reversed(messages[-1].content):
+                            if isinstance(content, TextContent):
+                                last_text_content = content
+                                break
+                        if last_text_content:
```
Comment (xingyaoww) [inline openhands/agenthub/codeact_agent/codeact_agent.py:491]: If we can't find the last_text_content, does it means that we will remove the new message?

File: openhands/controller/agent_controller.py (hunk 1 above)
Comment (enyst) [reply openhands/controller/agent_controller.py:339]: The failed tool calls are gathered in the [pending queue](https://github.com/All-Hands-AI/OpenHands/blob/409d132747bfcc9d77b5eb56a8936856a89b6df8/openhands/agenthub/codeact_agent/codeact_agent.py#L456) in the agent otherwise, and never get sent to the LLM. They are still in the stream, but the agent _never_ sees them, even though they appear real to the user (and to us, when sent in trajectories/logs etc).

Actually, I think we should give it an error observation. It appears to work, I'm still testing that. What do you think?

Comment (xingyaoww) [reply openhands/controller/agent_controller.py:339]: What does "failed tool calls" means? Is it a tool call that somehow never get executed?

Comment (enyst) [reply openhands/controller/agent_controller.py:339]: Ah, yes, sorry. When there's a runtime error like one raised [here](https://github.com/All-Hands-AI/OpenHands/blob/8488dd2a03db024ca6f39f0f30135b1156f4bf2c/openhands/runtime/impl/eventstream/eventstream_runtime.py#L442), it's possible it never gets an observation. Like this: 
- https://github.com/All-Hands-AI/OpenHands/issues/5601
- https://github.com/All-Hands-AI/OpenHands/issues/5602

Comment (xingyaoww) [reply openhands/controller/agent_controller.py:339]: ahh - then how would remove the tool call metadata here help though?

Comment (enyst) [reply openhands/controller/agent_controller.py]: Oh, I think you're right to question it - it should fail some asserts in the agent if it was removed. No, this is no good.

File: openhands/agenthub/codeact_agent/codeact_agent.py (hunk 2 above)
Comment (enyst) [reply openhands/agenthub/codeact_agent/codeact_agent.py:491]: Fixing!

File: openhands/controller/agent_controller.py (hunk 1 above)
Comment (enyst) [reply openhands/controller/agent_controller.py]: Here in `reset`, we still have the `self._pending_action`, and we know this action isn't actually going to finish normally. We can try to create on spot an error obs, linked to that action? 🤔 

I'm not sure there are many alternatives right now, I can think of others but they won't work... Does the agent really need to support multiple pending tool calls, when in fact we're processing them by response (ModelResponse)? And we [override](https://github.com/All-Hands-AI/OpenHands/blob/8488dd2a03db024ca6f39f0f30135b1156f4bf2c/openhands/agenthub/codeact_agent/codeact_agent.py#L172) the tool call to have one per ModelResponse. We're executing it, it doesn't actually wait for another?

Comment (xingyaoww) [reply openhands/controller/agent_controller.py]: >  We can try to create on spot an error obs, linked to that action? 🤔

I think this is probably a better way! This also preserve enough info so we can look into details when it happens again

File: openhands/agenthub/codeact_agent/codeact_agent.py (hunk 2 above)
Comment (enyst) [reply openhands/agenthub/codeact_agent/codeact_agent.py:491]: By the way, we probably shouldn't be doing this kind of thing at all anymore... Anthropic finally doesn't need it anymore, it accepts u/u/a/a. And litellm is doing it for providers that still need it.

Comment (xingyaoww) [reply openhands/agenthub/codeact_agent/codeact_agent.py:491]: Yeah! Let's remove it then - it makes our code much more ugly :(

**Comment by enyst** [comment]
@xingyaoww This is ready for review. Not tested with UI yet, but headless it works fine and looks ok.

File: openhands/controller/agent_controller.py [hunk 3]
```diff
@@ -334,6 +334,27 @@ async def _handle_message_action(self, action: MessageAction) -> None:
...
+                    and event.tool_call_metadata
+                    == self._pending_action.tool_call_metadata
+                ):
+                    found_observation = True
+                    break
+
+            # make a new ErrorObservation with the tool call metadata
+            if not found_observation:
+                obs = ErrorObservation(content='The action has not been executed.')
+                obs.tool_call_metadata = self._pending_action.tool_call_metadata
```
Comment (xingyaoww) [inline openhands/controller/agent_controller.py:353]: Does that means this `ErrorObservation` will trigger the execution of the same action or no? Will be nice if we can add a test for this..

File: openhands/llm/llm.py [hunk 4]
```diff
@@ -240,12 +245,20 @@ def wrapper(*args, **kwargs):
...
+                tool_calls: list[ChatCompletionMessageToolCall] = resp['choices'][0][
+                    'message'
+                ].get('tool_calls', [])
                 if tool_calls:
                     for tool_call in tool_calls:
-                        fn_name = tool_call.function.name
-                        fn_args = tool_call.function.arguments
-                        message_back += f'\nFunction call: {fn_name}({fn_args})'
+                        fn: Function | None = (
+                            tool_call.model_extra.get('function', None)
```
Comment (xingyaoww) [inline openhands/llm/llm.py:254]: This this the best practice to access function from `model_extra`?

**Review by xingyaoww — COMMENTED** [review]
Mostly LGTM! Just hoping to have a test -- maybe we can ask OpenHands to help here?

File: openhands/controller/agent_controller.py (hunk 3 above)
Comment (enyst) [reply openhands/controller/agent_controller.py:353]: It won't, because
1) next lines will make `self._pending_action = None` in the controller, so it won't keep waiting for an obs,
2) it's an obs with `tool_id`, so in the agent too, the pending tool call is now "complete"

On the contrary, leaving it pending is problematic 🤔

File: openhands/controller/agent_controller.py [hunk 5]
```diff
@@ -334,6 +334,27 @@ async def _handle_message_action(self, action: MessageAction) -> None:
...
+                    isinstance(event, Observation)
+                    and event.tool_call_metadata
+                    == self._pending_action.tool_call_metadata
+                ):
+                    found_observation = True
+                    break
+
+            # make a new ErrorObservation with the tool call metadata
+            if not found_observation:
+                obs = ErrorObservation(content='The action has not been executed.')
```
[the remark this answers is not in the document]
Comment (enyst) [reply openhands/controller/agent_controller.py:352]: Should we? The agent also adds `"[Error occurred in processing last action]"`
https://github.com/All-Hands-AI/OpenHands/blob/0dd919bacf10677f8885e78aef43259251334202/openhands/agenthub/codeact_agent/codeact_agent.py#L305

The agent is in STOPPED or in ERROR at this point. So with the UI, control is back to the user. What the agent should do might be different after the user message?

File: openhands/llm/llm.py (hunk 4 above)
Comment (enyst) [reply openhands/llm/llm.py:254]: Quite the contrary, I'd rather not rely on `model_extra` at all. Even though we do this for other things. The reason for this change is that it wasn't logging at all, getting function didn't work. Which is weird... Let me try again.

image

**Comment by openhands-agent** [comment]
Overview of Changes:
✓ Successfully implemented comprehensive unit tests for reset() function in agent_controller.py
✓ Added test coverage for all major scenarios and edge cases
✓ Properly implemented mocking and test infrastructure
✓ Verified core functionality and error handling

Status: All requested changes have been successfully completed with no remaining issues. The implementation provides thorough test coverage and addresses all key aspects of the reset() functionality.

File: openhands/llm/llm.py (hunk 4 above)
Comment (enyst) [reply openhands/llm/llm.py]: Reverted. litellm aims to be openai-compatible so it uses some magic to do it - and it works for me now

File: openhands/controller/agent_controller.py [hunk 6]
```diff
@@ -335,6 +335,27 @@ async def _handle_message_action(self, action: MessageAction) -> None:
     def _reset(self) -> None:
         """Resets the agent controller"""
 
+        # make sure there is an Observation with the tool call metadata to be recognized by the agent
```
Comment (enyst) [inline openhands/controller/agent_controller.py:338]:
```suggestion
        # make sure there is an Observation with the tool call metadata to be recognized by the agent
        # otherwise the pending action is found in history, but it's incomplete without an obs with tool result
```

File: openhands/controller/agent_controller.py (hunk 3 above)
Comment (enyst) [reply openhands/controller/agent_controller.py:355]: `openhands-agent` added some nice unit tests! They show the behavior here

File: openhands/controller/agent_controller.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands/controller/agent_controller.py:354]: Hmm if it is back to the user, maybe we can add something like "This is not the agent's fault, but rather runtime? Please consider to ask agent to try again"

And I wonder if we should just show this error and ask the agent itself to try again, without bothering the user, since this is really runtime's fault?

[the remark this answers is not in the document]
Comment (enyst) [reply openhands/controller/agent_controller.py:354]: > And I wonder if we should just show this error and ask the agent itself to try again, without bothering the user, since this is really runtime's fault?

I am a bit wary, because it feels like going back to what was messy at some point in time: errors that should stop things (exceptions) and errors that are normal agent history (ErrorObs) had become merged and interleaved, and the flow was quite a mess.

This code is a `reset`, and it can happen after a RuntimeError. And yet, because of tool use behavior, we are making an ErrorObservation this time... But then, we feel like we want the loop to continue? 🤔
- why didn't we make an ErrorObs upstream in the execution flow? Because this error was very bad and the agent cannot help with it, I think.
- this code is also executed when the *user* used the "Stop" UI button. Then it should definitely stay stopped.

We may need to give it some thought.

[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands/controller/agent_controller.py:354]: @rbren thoughts for this?

[the remark this answers is not in the document]
Comment (enyst) [reply openhands/controller/agent_controller.py:354]: This PR was in the back of my mind these days and I think that maybe the best place for this extra observation is not be here, in the controller, but rather in the agent or an agent helper (some kind of message history).

It might not even need to be a "real" obs in the stream, but handle the case "this action will never complete" (in tool use sense, it will always be "pending").

[the remark this answers is not in the document]
Comment (rbren) [reply openhands/controller/agent_controller.py:354]: This does seem reasonable. Not sure what the best place for it is, but nothing seems particularly better than this
